### PR TITLE
[canvaskit] Fall back to multi-context rendering on Firefox and Safari

### PR DESCRIPTION
**Pull request by harryterkelsen**
This PR refactors rasterization to support two different modes, single GL context mode backed by a single OffscreenCanvas (Chrome), and multi-context mode which uses many GL contexts and canvases.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by harryterkelsen** [comment]
Will fix goldens tomorrow

File: lib/web_ui/lib/src/engine/canvaskit/multi_surface_rasterizer.dart [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+import 'package:ui/src/engine.dart';
+import 'package:ui/ui.dart' as ui;
+
+/// A Rasterizer which uses one or many on-screen WebGL contexts to display the
+/// scene. This way of rendering is prone to bugs because there is a limit to
+/// how many WebGL contexts can be live at one time as well as bugs in sharing
+/// GL resources between the contexts. However, using [createImageBitmap] is
+/// currently very slow on Firefox and Safari browsers, so directly rendering
+/// to several
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/multi_surface_rasterizer.dart:13]: I'm listening... :)

File: lib/web_ui/lib/src/engine/canvaskit/offscreen_canvas_rasterizer.dart [hunk 2]
```diff
@@ -0,0 +1,63 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:ui/src/engine.dart';
+
+/// A [Rasterizer] that uses a single GL context in an OffscreenCanvas to do
+/// all the rendering. It transers bitmaps created in the OffscreenCanvas to
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/offscreen_canvas_rasterizer.dart:8]: typo: s/transers/transfers/

File: lib/web_ui/lib/src/engine/canvaskit/rasterizer.dart [hunk 3]
```diff
@@ -5,59 +5,111 @@
...
+  /// Disposes this rasterizer.
   void dispose() {
     viewEmbedder.dispose();
-    renderCanvasFactory.dispose();
+    overlayFactory.dispose();
   }
 }
+
+abstract class OverlayCanvas {
+  DomElement get htmlElement;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/rasterizer.dart:105]: nit: HTML is already implied by the type; instead consider using the name to convey the function of the element, such as "hostElement", "canvasElement", or "rootElement".

File: lib/web_ui/lib/src/engine/canvaskit/surface.dart [hunk 4]
```diff
@@ -47,11 +49,19 @@ class SurfaceFrame {
...
+  Surface({this.isRenderCanvas = false})
+      : useOffscreenCanvas =
+            Surface.offscreenCanvasSupported && !isRenderCanvas;
 
   CkSurface? _surface;
 
+  /// Whether or not to use an `OffscreenCanvas` to back this [Surface].
+  final bool useOffscreenCanvas;
+
+  /// If `true`, this [Surface] is used for rendering.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/surface.dart:62]: Does "render" here mean "display", i.e. it's shown to the user? Surfaces are always used for some kind of rendering, so it's unclear what it means for it to not be a render canvas.

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart [hunk 5]
```diff
@@ -1237,6 +1237,7 @@ class EnginePlatformDispatcher extends ui.PlatformDispatcher {
   /// Otherwise zones won't work properly.
   void invokeOnSemanticsAction(
       int nodeId, ui.SemanticsAction action, ByteData? args) {
+    print('invoking semantics action callback!');
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/platform_dispatcher.dart:1240]: Remove?

File: lib/web_ui/test/canvaskit/overlay_canvas_factory_test.dart [hunk 6]
```diff
@@ -13,11 +13,13 @@ void main() {
 }
 
 void testMain() {
-  group('$RenderCanvasFactory', () {
+  group('$OverlayCanvasFactory', () {
```
Comment (yjbanov) [inline lib/web_ui/test/canvaskit/overlay_canvas_factory_test.dart:16]: How do we test the two strategies? We want to make sure tests cover single and multi context modes.

**Review by yjbanov — CHANGES_REQUESTED** [review]

File: lib/web_ui/lib/src/engine/canvaskit/multi_surface_rasterizer.dart (hunk 1 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/multi_surface_rasterizer.dart]: Oops. done

File: lib/web_ui/lib/src/engine/canvaskit/offscreen_canvas_rasterizer.dart (hunk 2 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/offscreen_canvas_rasterizer.dart]: done

File: lib/web_ui/lib/src/engine/canvaskit/rasterizer.dart (hunk 3 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/rasterizer.dart]: done

File: lib/web_ui/lib/src/engine/canvaskit/surface.dart (hunk 4 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/surface.dart]: Done. Also, renamed `OverlayCanvas` to `DisplayCanvas` to make it clearer what these canvases are used for.

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart (hunk 5 above)
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart]: oops. leftover debugging print.

File: lib/web_ui/test/canvaskit/overlay_canvas_factory_test.dart (hunk 6 above)
Comment (harryterkelsen) [reply lib/web_ui/test/canvaskit/overlay_canvas_factory_test.dart]: The only difference between those modes is the factory function which creates the DisplayCanvas. This test is just for the functions of `DisplayCanvas` itself, so it should be agnostic to what kinds of `DisplayCanvas`es are actually made. I refactored the test a bit to clarify that.

**Review by harryterkelsen — COMMENTED** [review]
Thanks! PTAL

**Comment by harryterkelsen** [comment]
PTAL. I also added a test for the off-by-one blurriness issue in `surface_test.dart`.

**Review by yjbanov — APPROVED** [review]
lgtm

**Comment by parlough** [comment]
Thanks @harryterkelsen!! Hope this will end up being temporary 🤞

**Comment by tomekit** [comment]
Just a question, how do I check whether this landed in stable channel or perhaps if you know if this will end up somewhere in the 3.19.x?

**Comment by mreinfurt** [comment]
> Just a question, how do I check whether this landed in stable channel or perhaps if you know if this will end up somewhere in the 3.19.x?

Looking for the same information here. Would be awesome if this landed in a hotfix.

**Pull request merged**